server/src/models/timestamps.rs

```diff
@@ -0,0 +1,350 @@
+    let sampling_interval = (end_time - start_time) as usize / (length - 1);
+    let mut timestamp_builder = TimestampBuilder::new(length);
+
+    for timestamp in (start_time..=end_time).step_by(sampling_interval) {
+        timestamp_builder.append_value(timestamp);
+    }
+    timestamp_builder.finish()
+}
+
+/// Decompress all of a segment's timestamps, which for this segment are sampled
+/// at an irregular sampling interval, and thus compressed using Gorilla's
+/// compression method for timestamps.
+fn decompress_all_irregular_timestamps(
+    start_time: Timestamp,
+    end_time: Timestamp,
+    residual_timestamps: &[u8],
+) -> TimestampArray {
+    // TODO: remove the casts when refactoring the query engine to use unsigned.
+    // TODO: replace the pre-allocation when Gorilla is only used as a fallback.
+    // As the number of timestamps encoded in `residual_timestamps` is unknown,
+    // `timestamp_builder` cannot be perfectly pre-allocated. However, as the
+    // Gorilla model type is bounded by `compression::GORILLA_MAXIMUM_LENGTH`,
+    // this generally becomes the predominant length of the compressed segments.
+    let mut timestamp_builder = TimestampBuilder::new(compression::GORILLA_MAXIMUM_LENGTH);
+
+    // Add the first timestamp stored as `start_time` in the segment.
+    timestamp_builder.append_value(start_time);
+
+    // Remove the one bit used as a flag to specify that Gorilla is used.
+    let mut bits = BitReader::try_new(residual_timestamps).unwrap();
+    bits.read_bit();
+
+    // Decompress the second timestamp stored as a delta in 14 bits.
+    let mut last_delta = bits.read_bits(14);
+    let mut timestamp = start_time + last_delta as i64;
+    timestamp_builder.append_value(timestamp);
+
+    // Decompress the remaining residual timestamps.
+    while !bits.is_empty() {
+        // Read the next flag with the value of 0, 10, 110, 1110, or 1111.
+        let mut leading_one_bits = 0;
+        while leading_one_bits < 4 && !bits.is_empty() && bits.read_bit() {
+            leading_one_bits += 1;
+        }
+
+        // Any leftover bits in residual_timestamps are set to one. Thus, a
+        // sequence of one bits followed by fewer bits than specified by the
+        // flag means that all residual timestamps have been decompressed.
+        if leading_one_bits != 0 && bits.remaining_bits() < 7 {
+            break;
+        }
+
+        let delta = match leading_one_bits {
+            0 => last_delta,                                               // Flag is 0.
+            1 => read_decode_and_compute_delta(&mut bits, 7, last_delta),  // Flag is 10.
+            2 => read_decode_and_compute_delta(&mut bits, 9, last_delta),  // Flag is 110.
+            3 => read_decode_and_compute_delta(&mut bits, 12, last_delta), // Flag is 1110.
+            4 => last_delta + bits.read_bits(32),                          // Flag is 1111.
+            _ => panic!("Unknown encoding of timestamps."),
```

Comment:
Shouldn't the flag be printed to help debugging then?

Reply:
Unfortunately the flag is no longer available on [Line 251](https://github.com/ModelarData/ModelarDB-RS/blob/479965545208fd5755e39b83e32bd873daed0631/server/src/models/timestamps.rs#L251) as the flag bits read from  the compressed representation is discarded after being read on [Line 234](https://github.com/ModelarData/ModelarDB-RS/blob/479965545208fd5755e39b83e32bd873daed0631/server/src/models/timestamps.rs#L234). Also as far as I can tell it is not possible to read a flag that is not handled by the `match` statement since the loop on [Line 234](https://github.com/ModelarData/ModelarDB-RS/blob/479965545208fd5755e39b83e32bd873daed0631/server/src/models/timestamps.rs#L234) stops after reading four bits due to `leading_one_bits < 4` and immediately if there are no more bits or it reads a zero bit due to `! bits.is_empty() && bits.read_bit()`. However, Rust is very pedantic when it comes to `match` statements and requires that all possible matches are handled, otherwise the following error is emitted `patterns i32::MIN..=-1_i32 and 5_i32..=i32::MAX not covered`. I have changed the match arm to the following to provide as much debugging information as possible `_ => panic!("Unknown timestamp encoding with {} leading one bits.", leading_one_bits),`.